polkadot/xcm/xcm-simulator/example/src/lib.rs

```diff
@@ -649,4 +649,23 @@ mod tests {
 			);
 		});
 	}
+
+	#[test]
+	fn builder_pattern_works() {
+		let asset: MultiAsset = (Here, 100u128).into();
+		let beneficiary: MultiLocation = AccountId32 { id: [0u8; 32], network: None }.into();
+		let message: Xcm<()> = Xcm::builder()
+			.withdraw_asset(asset.clone().into())
+			.buy_execution(asset.clone(), Unlimited)
+			.deposit_asset(asset.clone().into(), beneficiary)
+			.build();
```

Comment:
The order of the methods here matters, right?
My intuitive to general builder model is that the order should not matter at all.
I suggest use something like,
```
Xcm::new()
  .withdraw_asset(...)
  .buy_execution(..)
  .deposit_asset(...)
```
If using this chained method syntax, we probably don't need to introduce a new macro? (there are already a lot of macros in the codebase.)

Reply:
The derive macro acts as a convenient way to create methods that are very similar to the declaration of the variants in the `Instruction` enum.

I also find it quite absurd to suggest that the order shouldn't matter. How could it be different? Every XCM instruction order matters in an XCM program. The argument that you're making with your example is simply whether or not we need to call the `builder` method before chaining instruction methods, however this has been a [Rust conventional pattern to create a `builder` method](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html). Is there a good reason why we shouldn't do this, other than the mistaken assumption that it would remove the derive macro?

Reply:
I think what he means is in the traditional builder pattern, the order of methods does not matter, and aims to make this clear by using `new()` instead of the usual `builder()`.
I have nothing against doing this change. It's a few characters less.

Reply:
What is true is we do need the macro, although we could change its name from `Builder` to something else, like `VariantsToMethods` but I'd argue it's not that off from the builder pattern, just the order thing, which we would easily deal with via documentation/communication

Reply:
> the order of methods does not matter

How does this make _any_ sense? Would the order of XCM instructions not matter? Why is this even being suggested as something viable?

Reply:
> in the traditional builder pattern, the order of methods does not matter

Yet we are using the builder pattern to build XCM, where the instruction order matters. If @kaichaosun means what you described, then it would seem it is a suggestion coming from a lack of understanding of what XCM instructions and programs are.

Reply:
> But then the `build` at the end feels kind of weird to me.

If the chained method makes sense, we don't need `build` any more.

> The derive macro acts as a convenient way to create methods that are very similar to the declaration of the variants in the `Instruction` enum.

Yeah it's convenient for writers, but it's not convenient for the readers/users, and this interface is meant to be used by runtime devs, using macro here seems take the wrong balance. Generally I'm not fan of using macros if the logic is business heavy, especially in user interface.

BTW, not sure if it makes sense, but I'd like to see something like this in the future with xcm,
```
Xcm.new()
   .withdraw(...),
   .fuel(...)
   .deposit(...)
```
Which feels just like using `map`, `filter` on iterator. There is `iter()` and `collect()` to convert between iterator and vector, you probably know better identifiers what should be used in Xcm if such abstraction is needed when multi Xcm exists. Before that, `new` seems enough? Also, for better UX, it may not be necessarily 1 to 1 map between such convenient method and the xcm instruction variant.

Note, I'm not a heavy XCM user, my suggestion is based on the my general programing experience, and what potentially can provide a better dev experience.

Reply:
I'm sorry, but you're not making much sense. Users do not dig deep into the macro code to look at the function body of e.g. `withdraw_asset`, the only API they care about is just the function. Hence readability of the macro code is not important for DevEx.

Reply:
> Also, for better UX, it may not be necessarily 1 to 1 map between such convenient method and the xcm instruction variant.

You are basically saying that we should build yet another abstraction layer on top of XCM, so that users can use XCM without actually knowing about XCM. This is a non-goal; the goal of XCM isn't to make it easier for end-users to craft XCM programs, but rather facilitate developers (especially parachain developers) to innovate upon cross-chain interactions. What you're describing is also no different from a macro, from the perspective that they're both trying to hide complexity, when the goal here _isn't_ that, but simply to make XCM nicer for developers to work with.

Reply:
I think using different types, so `XcmBuilder` accessed via `Xcm::builder()` that then has `.build()` to go back to `Xcm` is a good thing.
It lets us add some nice limitations when building the message and then making sure that when calling `build()` it's valid, as we deem it.